README.md
# dart_ssss 

[![Build Status](https://travis-ci.org/gching/dart_ssss.svg?branch=master)](https://travis-ci.org/gching/dart_ssss)
[![Coverage Status](https://coveralls.io/repos/github/gching/dart_ssss/badge.svg?branch=master)](https://coveralls.io/github/gching/dart_ssss?branch=master)

A Dart library that implements [Shamir's Secret Sharing Scheme](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing) 
over [GF(256)](http://www.cs.utsa.edu/~wagner/laws/FFM.html).

Given a secret in bytes, this package can assist in splitting up the secret into `n` parts and would be able to 
reconstruct the secret in bytes given a `t` threshold.

I would highly recommend you look at the the wikipedia page above to learn about the workings of Shamir's Secret Sharing
Scheme.

## Getting Started

To use `dart_ssss`, add it to your dependencies in `pubspec.yaml`:

```
dependencies:
    dart_ssss: any
```

Then import the proper package:

```
import 'package:dart_ssss/dart_ssss.dart';
```


## Usage 

To generate shares / parts for a secret of bytes:

```
void main() {
    
    List<int> secretInByteValues = [...];
    
    SecretScheme ss = new SecretScheme(numberOfPartsToGenerate, thresholdForJoining);
    
    Map<int, List<int>> shares = ss.createShares(secretInByteValues);
    
    ...
      
     
```

To combine shares to regenerate a secret:

```
    List<int> recombinedSecretInBytes = ss.combineShares(shares);
}
```

## Tests

To run tests on this package:

```
    pub run test 
```

in the package directory.


## Limitations / Caveats

Given that the package is implemented over GF(256), it means that the max number of shares / parts it can generate is 
`255`. With that said, it is not advisable to generate `255` shares / parts as an adversary can 
[forge a secret](https://crypto.stackexchange.com/questions/54578/how-to-forge-a-shamir-secret-share).    

Shamir's Secret Sharing Scheme does not guarantee integrity, so if you are looking for something where you can verify
shares or the resulting secret, this package is not for you.

## License

Copyright 2018 Gavin Ching

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



CHANGELOG.md
# Changelog

## 0.0.1

- Initial version with all functionalities implemented.

## 0.0.2

- Add coverage analysis through Coveralls (`tool/coveralls.sh`) and have more tests for higher coverage.
- Better readme.

## 0.0.3

- Fix issue of possibly return a `0` x coordinate as a share.
- Add in branches of code to ensure that it will be tested later on (lowering coverage).

## 0.0.4

- Remove random byte method from ByteHelper and create ByteRandom singleton to be used to generate random bytes.
- Add mocks to tests to properly test edge cases involving specific random bytes.

test/test_all.dart
/**
 * test_all.dart
 *
 * Copyright 2018 Gavin Ching
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR I
 * MPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

import 'byte_polynomial_test.dart' as byte_polynomial;
import 'gf_256_test.dart' as gf_256;
import 'lagrange_interpolation_test.dart' as lagrange_interpolation;
import 'secret_scheme_test.dart' as secret_scheme;
import 'utils/byte_helper_test.dart' as byte_helper;
import 'utils/byte_random_test.dart' as byte_random;

void main() {
  byte_polynomial.main();
  gf_256.main();
  lagrange_interpolation.main();
  secret_scheme.main();
  byte_helper.main();
  byte_random.main();
}


test/gf_256_test.dart
/**
 * GF256Test.dart
 *
 * Copyright 2018 Gavin Ching
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR I
 * MPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

import 'package:test/test.dart';
import 'package:dart_ssss/src/gf_256.dart';

void main() {
  group('Adding and Subtracting', () {
    test('Adding two zeroes should return zero', () {
      expect(GF256.add(0, 0), equals(0));
    });

    test('Adding / Subtracting zero and one should return one', () {
      expect(GF256.add(0, 1), equals(1));
      expect(GF256.sub(0, 1), equals(1));
    });

    test('Adding / Subtracting 255 and 255 should return zero', () {
      expect(GF256.add(255, 255), equals(0));
      expect(GF256.sub(255, 255), equals(0));
    });
  });

  group('Multiplication', () {
    test('Multiplying two zeroes should return zero', () {
      expect(GF256.multiply(0, 0), equals(0));
    });

    test('Multiplying one with one should return one', () {
      expect(GF256.multiply(1, 1), equals(1));
    });

    test('Multiplying with multiple operands should return the correct results',
        () {
      expect(GF256.multiply(0xb6, 0x53), equals(0x36));
    });
  });

  group('Division', () {
    test('Dividing should return the correct results', () {
      expect(GF256.divide(90, 21), equals(189));
      expect(GF256.divide(6, 55), equals(151));
      expect(GF256.divide(22, 192), equals(138));
      expect(GF256.divide(0, 192), equals(0));
    });
  });

  group('Throw on not byte', () {
    test('Should throw if the value is not a byte', () {
      expect(() => GF256.throwIfNotUnsignedByte(-1), throwsArgumentError);
    });
  });
}


test/utils/byte_helper.dart
/**
 * byte_helper.dart
 *
 * Copyright 2018 Gavin Ching
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR I
 * MPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

import 'package:dart_ssss/src/utils/byte_helper.dart';
import 'package:test/test.dart';

void main() {
  group('Byte Checking', () {
    test('Should return the proper booleans for byte checking', () {
      expect(ByteHelper.isByte(0), isTrue);
      expect(ByteHelper.isNotByte(0), isFalse);

      expect(ByteHelper.isByte(-1), isFalse);
      expect(ByteHelper.isNotByte(-1), isTrue);
    });

    test('Should return the proper boolean values for byte checking on a list',
        () {
      List<int> bytes = [1, 2, 3, 4, 5];
      List<int> notBytes = [-1, 5, -1, 256];

      expect(ByteHelper.isListAllBytes(bytes), isTrue);
      expect(ByteHelper.isListAllBytes(notBytes), isFalse);
    });
  });
}


lib/dart_ssss.dart
/**
 * dart_ssss.dart
 *
 * Copyright 2018 Gavin Ching
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR I
 * MPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

/// # dart_ssss
///
/// ## Getting Started
///
/// To generate shares / parts for a secret of bytes:
///
/// ```
///   void main() {
///
///     List<int> secretInByteValues = [...];
///
///     SecretScheme ss = new SecretScheme(numberOfPartsToGenerate, thresholdForJoining);
///
///     Map<int, List<int>> shares = ss.createShares(secretInByteValues);
///
///     ...
///
/// ```
///
/// To combine shares to regenerate a secret:
///
/// ```
///
///     List<int> recombinedSecretInBytes = ss.combineShares(shares);
///
///   }
///
/// ```
library dart_ssss;

export 'src/secret_scheme.dart';


tools/coveralls.sh
#!/bin/bash
# Copyright 2015, Google Inc. All rights reserved.
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:

#    * Redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer.
#    * Redistributions in binary form must reproduce the above
# copyright notice, this list of conditions and the following disclaimer
# in the documentation and/or other materials provided with the
# distribution.

#    * Neither the name of Google Inc. nor the names of its
# contributors may be used to endorse or promote products derived from
# this software without specific prior written permission.

# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

# Fast fail the script on failures.
set -e

# Install dart_coveralls; gather and send coverage data.
if [ "$COVERALLS_TOKEN" ] && [ "$TRAVIS_DART_VERSION" = "stable" ] && [ "$TRAVIS_DART_TEST" = true ]; then
  pub global activate dart_coveralls
  pub global run dart_coveralls report \
    --token $COVERALLS_TOKEN \
    --retry 2 \
    --exclude-test-files \
    test/test_all.dart
fi

lib/src/utils/byte_helper.dart
/**
 * byte_helper.dart
 *
 * Copyright 2018 Gavin Ching
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR I
 * MPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

/**
 * Class that provides methods to help with bytes.
 */
class ByteHelper {
  /**
   * Returns true if all values in the provided list is representable as an
   * unsigned byte value (0-255).
   */
  static bool isListAllBytes(List<int> vals) {
    // If it is null, return false.
    if (vals == null) {
      return false;
    }

    // If any of the values in the list is not a byte, return false.
    if (vals.any(isNotByte)) {
      return false;
    }

    // All values are bytes, return true.
    return true;
  }

  /**
   * Returns true if the val is representable as an unsigned byte value (0-255).
   */
  static bool isByte(int val) {
    return val >= 0 && val <= 255;
  }

  /**
   * Returns true if the val is NOT representable as an unsigned byte value
   * (0-255).
   */
  static bool isNotByte(int val) {
    return !isByte(val);
  }
}


lib/src/gf_256.dart
/**
 * gf_256.dart
 *
 * Copyright 2018 Gavin Ching
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR I
 * MPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

import 'utils/byte_helper.dart';

/**
 * Implementation of operations on finite field GF(2^8) or GF(256) to be
 * utilized with the SSSS. To leare more about GF(256) or why it was selected,
 * please refer to:
 *
 * https://research.swtch.com/field.
 *
 * Important thing to note is that because we have selected a finite field
 * GF(256), it means that we are limited to number of shares cannot be greater
 * then 255 (where one coordinate is saved to decipher the secret).
 */
class GF256 {
  static final List<int> LOG3 = [
    0xff,
    0x00,
    0x19,
    0x01,
    0x32,
    0x02,
    0x1a,
    0xc6,
    0x4b,
    0xc7,
    0x1b,
    0x68,
    0x33,
    0xee,
    0xdf,
    0x03,
    0x64,
    0x04,
    0xe0,
    0x0e,
    0x34,
    0x8d,
    0x81,
    0xef,
    0x4c,
    0x71,
    0x08,
    0xc8,
    0xf8,
    0x69,
    0x1c,
    0xc1,
    0x7d,
    0xc2,
    0x1d,
    0xb5,
    0xf9,
    0xb9,
    0x27,
    0x6a,
    0x4d,
    0xe4,
    0xa6,
    0x72,
    0x9a,
    0xc9,
    0x09,
    0x78,
    0x65,
    0x2f,
    0x8a,
    0x05,
    0x21,
    0x0f,
    0xe1,
    0x24,
    0x12,
    0xf0,
    0x82,
    0x45,
    0x35,
    0x93,
    0xda,
    0x8e,
    0x96,
    0x8f,
    0xdb,
    0xbd,
    0x36,
    0xd0,
    0xce,
    0x94,
    0x13,
    0x5c,
    0xd2,
    0xf1,
    0x40,
    0x46,
    0x83,
    0x38,
    0x66,
    0xdd,
    0xfd,
    0x30,
    0xbf,
    0x06,
    0x8b,
    0x62,
    0xb3,
    0x25,
    0xe2,
    0x98,
    0x22,
    0x88,
    0x91,
    0x10,
    0x7e,
    0x6e,
    0x48,
    0xc3,
    0xa3,
    0xb6,
    0x1e,
    0x42,
    0x3a,
    0x6b,
    0x28,
    0x54,
    0xfa,
    0x85,
    0x3d,
    0xba,
    0x2b,
    0x79,
    0x0a,
    0x15,
    0x9b,
    0x9f,
    0x5e,
    0xca,
    0x4e,
    0xd4,
    0xac,
    0xe5,
    0xf3,
    0x73,
    0xa7,
    0x57,
    0xaf,
    0x58,
    0xa8,
    0x50,
    0xf4,
    0xea,
    0xd6,
    0x74,
    0x4f,
    0xae,
    0xe9,
    0xd5,
    0xe7,
    0xe6,
    0xad,
    0xe8,
    0x2c,
    0xd7,
    0x75,
    0x7a,
    0xeb,
    0x16,
    0x0b,
    0xf5,
    0x59,
    0xcb,
    0x5f,
    0xb0,
    0x9c,
    0xa9,
    0x51,
    0xa0,
    0x7f,
    0x0c,
    0xf6,
    0x6f,
    0x17,
    0xc4,
    0x49,
    0xec,
    0xd8,
    0x43,
    0x1f,
    0x2d,
    0xa4,
    0x76,
    0x7b,
    0xb7,
    0xcc,
    0xbb,
    0x3e,
    0x5a,
    0xfb,
    0x60,
    0xb1,
    0x86,
    0x3b,
    0x52,
    0xa1,
    0x6c,
    0xaa,
    0x55,
    0x29,
    0x9d,
    0x97,
    0xb2,
    0x87,
    0x90,
    0x61,
    0xbe,
    0xdc,
    0xfc,
    0xbc,
    0x95,
    0xcf,
    0xcd,
    0x37,
    0x3f,
    0x5b,
    0xd1,
    0x53,
    0x39,
    0x84,
    0x3c,
    0x41,
    0xa2,
    0x6d,
    0x47,
    0x14,
    0x2a,
    0x9e,
    0x5d,
    0x56,
    0xf2,
    0xd3,
    0xab,
    0x44,
    0x11,
    0x92,
    0xd9,
    0x23,
    0x20,
    0x2e,
    0x89,
    0xb4,
    0x7c,
    0xb8,
    0x26,
    0x77,
    0x99,
    0xe3,
    0xa5,
    0x67,
    0x4a,
    0xed,
    0xde,
    0xc5,
    0x31,
    0xfe,
    0x18,
    0x0d,
    0x63,
    0x8c,
    0x80,
    0xc0,
    0xf7,
    0x70,
    0x07
  ];

  static final List<int> EXP3 = [
    0x01,
    0x03,
    0x05,
    0x0f,
    0x11,
    0x33,
    0x55,
    0xff,
    0x1a,
    0x2e,
    0x72,
    0x96,
    0xa1,
    0xf8,
    0x13,
    0x35,
    0x5f,
    0xe1,
    0x38,
    0x48,
    0xd8,
    0x73,
    0x95,
    0xa4,
    0xf7,
    0x02,
    0x06,
    0x0a,
    0x1e,
    0x22,
    0x66,
    0xaa,
    0xe5,
    0x34,
    0x5c,
    0xe4,
    0x37,
    0x59,
    0xeb,
    0x26,
    0x6a,
    0xbe,
    0xd9,
    0x70,
    0x90,
    0xab,
    0xe6,
    0x31,
    0x53,
    0xf5,
    0x04,
    0x0c,
    0x14,
    0x3c,
    0x44,
    0xcc,
    0x4f,
    0xd1,
    0x68,
    0xb8,
    0xd3,
    0x6e,
    0xb2,
    0xcd,
    0x4c,
    0xd4,
    0x67,
    0xa9,
    0xe0,
    0x3b,
    0x4d,
    0xd7,
    0x62,
    0xa6,
    0xf1,
    0x08,
    0x18,
    0x28,
    0x78,
    0x88,
    0x83,
    0x9e,
    0xb9,
    0xd0,
    0x6b,
    0xbd,
    0xdc,
    0x7f,
    0x81,
    0x98,
    0xb3,
    0xce,
    0x49,
    0xdb,
    0x76,
    0x9a,
    0xb5,
    0xc4,
    0x57,
    0xf9,
    0x10,
    0x30,
    0x50,
    0xf0,
    0x0b,
    0x1d,
    0x27,
    0x69,
    0xbb,
    0xd6,
    0x61,
    0xa3,
    0xfe,
    0x19,
    0x2b,
    0x7d,
    0x87,
    0x92,
    0xad,
    0xec,
    0x2f,
    0x71,
    0x93,
    0xae,
    0xe9,
    0x20,
    0x60,
    0xa0,
    0xfb,
    0x16,
    0x3a,
    0x4e,
    0xd2,
    0x6d,
    0xb7,
    0xc2,
    0x5d,
    0xe7,
    0x32,
    0x56,
    0xfa,
    0x15,
    0x3f,
    0x41,
    0xc3,
    0x5e,
    0xe2,
    0x3d,
    0x47,
    0xc9,
    0x40,
    0xc0,
    0x5b,
    0xed,
    0x2c,
    0x74,
    0x9c,
    0xbf,
    0xda,
    0x75,
    0x9f,
    0xba,
    0xd5,
    0x64,
    0xac,
    0xef,
    0x2a,
    0x7e,
    0x82,
    0x9d,
    0xbc,
    0xdf,
    0x7a,
    0x8e,
    0x89,
    0x80,
    0x9b,
    0xb6,
    0xc1,
    0x58,
    0xe8,
    0x23,
    0x65,
    0xaf,
    0xea,
    0x25,
    0x6f,
    0xb1,
    0xc8,
    0x43,
    0xc5,
    0x54,
    0xfc,
    0x1f,
    0x21,
    0x63,
    0xa5,
    0xf4,
    0x07,
    0x09,
    0x1b,
    0x2d,
    0x77,
    0x99,
    0xb0,
    0xcb,
    0x46,
    0xca,
    0x45,
    0xcf,
    0x4a,
    0xde,
    0x79,
    0x8b,
    0x86,
    0x91,
    0xa8,
    0xe3,
    0x3e,
    0x42,
    0xc6,
    0x51,
    0xf3,
    0x0e,
    0x12,
    0x36,
    0x5a,
    0xee,
    0x29,
    0x7b,
    0x8d,
    0x8c,
    0x8f,
    0x8a,
    0x85,
    0x94,
    0xa7,
    0xf2,
    0x0d,
    0x17,
    0x39,
    0x4b,
    0xdd,
    0x7c,
    0x84,
    0x97,
    0xa2,
    0xfd,
    0x1c,
    0x24,
    0x6c,
    0xb4,
    0xc7,
    0x52,
    0xf6,
    0x01,
    0x03,
    0x05,
    0x0f,
    0x11,
    0x33,
    0x55,
    0xff,
    0x1a,
    0x2e,
    0x72,
    0x96,
    0xa1,
    0xf8,
    0x13,
    0x35,
    0x5f,
    0xe1,
    0x38,
    0x48,
    0xd8,
    0x73,
    0x95,
    0xa4,
    0xf7,
    0x02,
    0x06,
    0x0a,
    0x1e,
    0x22,
    0x66,
    0xaa,
    0xe5,
    0x34,
    0x5c,
    0xe4,
    0x37,
    0x59,
    0xeb,
    0x26,
    0x6a,
    0xbe,
    0xd9,
    0x70,
    0x90,
    0xab,
    0xe6,
    0x31,
    0x53,
    0xf5,
    0x04,
    0x0c,
    0x14,
    0x3c,
    0x44,
    0xcc,
    0x4f,
    0xd1,
    0x68,
    0xb8,
    0xd3,
    0x6e,
    0xb2,
    0xcd,
    0x4c,
    0xd4,
    0x67,
    0xa9,
    0xe0,
    0x3b,
    0x4d,
    0xd7,
    0x62,
    0xa6,
    0xf1,
    0x08,
    0x18,
    0x28,
    0x78,
    0x88,
    0x83,
    0x9e,
    0xb9,
    0xd0,
    0x6b,
    0xbd,
    0xdc,
    0x7f,
    0x81,
    0x98,
    0xb3,
    0xce,
    0x49,
    0xdb,
    0x76,
    0x9a,
    0xb5,
    0xc4,
    0x57,
    0xf9,
    0x10,
    0x30,
    0x50,
    0xf0,
    0x0b,
    0x1d,
    0x27,
    0x69,
    0xbb,
    0xd6,
    0x61,
    0xa3,
    0xfe,
    0x19,
    0x2b,
    0x7d,
    0x87,
    0x92,
    0xad,
    0xec,
    0x2f,
    0x71,
    0x93,
    0xae,
    0xe9,
    0x20,
    0x60,
    0xa0,
    0xfb,
    0x16,
    0x3a,
    0x4e,
    0xd2,
    0x6d,
    0xb7,
    0xc2,
    0x5d,
    0xe7,
    0x32,
    0x56,
    0xfa,
    0x15,
    0x3f,
    0x41,
    0xc3,
    0x5e,
    0xe2,
    0x3d,
    0x47,
    0xc9,
    0x40,
    0xc0,
    0x5b,
    0xed,
    0x2c,
    0x74,
    0x9c,
    0xbf,
    0xda,
    0x75,
    0x9f,
    0xba,
    0xd5,
    0x64,
    0xac,
    0xef,
    0x2a,
    0x7e,
    0x82,
    0x9d,
    0xbc,
    0xdf,
    0x7a,
    0x8e,
    0x89,
    0x80,
    0x9b,
    0xb6,
    0xc1,
    0x58,
    0xe8,
    0x23,
    0x65,
    0xaf,
    0xea,
    0x25,
    0x6f,
    0xb1,
    0xc8,
    0x43,
    0xc5,
    0x54,
    0xfc,
    0x1f,
    0x21,
    0x63,
    0xa5,
    0xf4,
    0x07,
    0x09,
    0x1b,
    0x2d,
    0x77,
    0x99,
    0xb0,
    0xcb,
    0x46,
    0xca,
    0x45,
    0xcf,
    0x4a,
    0xde,
    0x79,
    0x8b,
    0x86,
    0x91,
    0xa8,
    0xe3,
    0x3e,
    0x42,
    0xc6,
    0x51,
    0xf3,
    0x0e,
    0x12,
    0x36,
    0x5a,
    0xee,
    0x29,
    0x7b,
    0x8d,
    0x8c,
    0x8f,
    0x8a,
    0x85,
    0x94,
    0xa7,
    0xf2,
    0x0d,
    0x17,
    0x39,
    0x4b,
    0xdd,
    0x7c,
    0x84,
    0x97,
    0xa2,
    0xfd,
    0x1c,
    0x24,
    0x6c,
    0xb4,
    0xc7,
    0x52,
    0xf6,
    0x00,
    0x00
  ];

  /**
   * Adds the left operand and right operand in GF(256) which is just the
   * bitwise XOR.
   */
  static int add(int leftOp, int rightOp) {
    throwIfNotUnsignedByte(leftOp);
    throwIfNotUnsignedByte(rightOp);

    // Perform bitwise XOR.
    return leftOp ^ rightOp;
  }

  /**
   * Subtracts the left operand and right operand in GF(256) which is add.
   */
  static int sub(int leftOp, int rightOp) => add(leftOp, rightOp);

  /**
   * Multplies the left operand and right operand in GF(256) which utilizes
   * exponential and log lookups given a generator of 0x03 (x + 1 polynomial).
   *
   * To learn more about this method:
   *
   * https://en.wikipedia.org/wiki/Finite_field_arithmetic#Implementation_tricks
   *
   */
  static int multiply(int leftOp, int rightOp) {
    throwIfNotUnsignedByte(leftOp);
    throwIfNotUnsignedByte(rightOp);

    // In the case that one of them is zero.
    if (leftOp == 0 || rightOp == 0) {
      return 0;
    }
    // Find the log3 of both the left operand and right operand
    // add them together and 3 to the power of the the addition results.
    //
    // results = 3^(log3(leftOp) + log3(rightOp))

    return EXP3[LOG3[leftOp] + LOG3[rightOp]];
  }

  /**
   * Does division given the dividend and divsor (dividend / divisor).
   * which utilizes lookups given a generator of 0x03 (x+1 polynomial).
   */
  static int divide(int dividend, int divisor) {
    throwIfNotUnsignedByte(dividend);
    throwIfNotUnsignedByte(divisor);

    // 0 divided by anything is 0.
    if (dividend == 0) {
      return 0;
    }

    // We calculate the multiplicative inverse which is defined by
    // inv = 3^(255 - log3(divisor))
    // Then we multiply the inv with the dividend which is the division results
    // of GF(256)
    return multiply(dividend, EXP3[255 - LOG3[divisor]]);
  }

  /**
   * Throws a byte exception error if the value is not a byte.
   */
  static void throwIfNotUnsignedByte(int possibleVal) {
    if (ByteHelper.isNotByte(possibleVal)) {
      throw new ArgumentError.value(possibleVal);
    }
  }
}
